USB: Add MS OS Descriptors version 1 support

finikorg opened the pull request.

finikorg committed:
usb: webusb: Use struct string_desc instead of char array

Use proper structures for dealing with MS OS v1 string descriptors.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Cleanup code style for usb_device

Make usb_device code consistent with the rest of the code.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Correct include path

This makes possible to include headers from subsys/usb

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: osdesc: Add MS OS Descriptors version 1 support

Add support for MS OS Descriptors. The patch structures already
existing code for webusb to the single place to be easy added to other
places.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: rndis: Add MS OS v1 descriptors to RNDIS

This should allow automatic driver loading in MS Windows.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: rndis: Set subCompatibleID parameter for exact MS driver

Set subCompatibleID value according to "C:\Windows\INF\rndiscmp.inf"
file. The exact matching part is:
"USB\MS_COMP_RNDIS&MS_SUBCOMP_5162001"

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@IvanSanchez

finikorg committed:
usb: tests: Add unit tests for MS OS Descriptors testing

Add test for testing MS OS String Descriptor version 1.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: tests: Add testing os_desc feature

Add testing usb_handle_os_desc_feature()

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

carlescufi commented:
@jfischer-phytec-iot @pawelzadrozniak any feedback on this one?

jfischer-phytec-iot commented:
will do, just have to find time

jfischer-phytec-iot commented:
@finikorg please rebase

finikorg commented:
@jfischer-phytec-iot rebased

jfischer-phytec-iot commented on subsys/usb/usb_device.c:
```diff
@@ -898,12 +922,19 @@ int usb_set_config(struct usb_cfg_data *config)
-	if (config->interface.vendor_handler) {
+
+	/* register vendor request handler */
+	if (config->interface.vendor_handler || usb_os_desc_enabled()) {
```
There is already a wrapper for composite configuration:
https://github.com/zephyrproject-rtos/zephyr/pull/7916/commits/b09e5dcb0e56f5ca6800938fdb27a91dbc194e8b#diff-609952c961f3c77cb7811bedee4bfd2cL1388

That also needs to be changed:
https://github.com/zephyrproject-rtos/zephyr/pull/7916/commits/b09e5dcb0e56f5ca6800938fdb27a91dbc194e8b#diff-609952c961f3c77cb7811bedee4bfd2cR1494

We can simplify it all and only use usb_composite_init for initialization without explicitly turning on CONFIG_USB_COMPOSITE_DEVICE, but first interface descriptors in samples webusb and wpanusb have to be reworked.

carlescufi replied:
@finikorg any comments on this? I think this makes sense

finikorg replied:
@jfischer-phytec-iot @carlescufi vendor_handler() in usb_devices.c does not handle request correctly. It expects that setup->wIndex indicates interface number, which is not the case for some (maybe all) of requests. 
"Microsoft OS 2.0 Descriptors Specification":
```
wIndex 0x7 for MS_OS_20_DESCRIPTOR_INDEX.
wIndex 0x8 for MS_OS_20_SET_ALT_ENUMERATION
...
etc
```

finikorg replied:
@jfischer-phytec-iot Added patch #afe05be

finikorg replied:
@jfischer-phytec-iot Please have a look at 96a9583

finikorg committed:
usb: composite: Add handling osdesc feature

Add handling osdesc feature also for composite configuration.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: rndis: Use RFC 7042 Doc value for Host MAC

Use similar RFC 7042 Documentation value like it is in CDC ECM.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Refactor vendor_handler

Since vendor commands cannot assume that wIndex keeps interface number
we change logic trying to call every vendor handler which should check
input parameters.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-phytec-iot commented on subsys/usb/os_desc.c:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	if (GET_DESC_TYPE(setup->wValue) == DESC_STRING &&
+	    GET_DESC_INDEX(setup->wValue) == 0xEE) {
```
Can 0xEE be replaced by a macro or otherwise documented?

finikorg replied:
This is possible, I will add documentation block.

finikorg committed:
usb: osdesc: Use definition for string descriptor index

Use definition and add description block.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>
